src/Analyser/MutatingScope.php

```diff
@@ -5721,4 +5721,27 @@ public function getIterableValueType(Type $iteratee): Type
 		return $iteratee->getIterableValueType();
 	}
 
+	public function getPhpVersion(): PhpVersion
+	{
+		$versionId = $this->getType(new ConstFetch(new Name('PHP_VERSION_ID')));
+		if ($versionId instanceof IntegerRangeType && $versionId->getMin() !== null) {
+			return new PhpVersion($versionId->getMin(), PhpVersion::SOURCE_RUNTIME);
+		}
+
+		$scalars = $versionId->getConstantScalarValues();
+		if ($scalars !== []) {
+			$ints = [];
+			foreach($scalars as $scalar) {
+				if (!is_int($scalar)) {
+					throw new ShouldNotHappenException();
+				}
+				$ints[] = $scalar;
```

Comment:
Looking at this piece of code I realized it's time to do an idea I had for a long time. Please follow my train of thought (while I'm typing this out on a train to Prague :))

Often there's not a single PHP version, but a range of them so returning `bool` from a method like this isn't exactly right:

```php
	public function supportsNullCoalesceAssign(): bool
	{
		return $this->versionId >= 70400;
	}
```

If we're analysing against both PHP 7.3 and 7.4+, we MIGHT support null coalesce assign.

If we're analysing only against PHP 7.3, we do NOT support null coalesce assign.

If we're analysing only against PHP 7.4+, we DO support null coalesce assign.

Sounds familiar? :) We should return TrinaryLogic instead of `bool`.

But I don't want the headache of BC breaks. Adding a new method about PHP version to Scope allows us to return something else than `PhpVersion`. A new unrelated class called something like `TrinaryPhpVersion` or `PhpVersions`. And we do not need to add all methods from PhpVersion on it. We can add them slowly as needed.

So in this PR this new class could have just `producesWarningForFinalPrivateMethods(): TrinaryLogic` correctly computed for all various scenarious of what `PHP_VERSION_ID` can be in that place.

And the rule would correctly handle yes/no/maybe along with some tests.

The constructor of this class could just accept a list of supported PHP versions like `80000, 80100, 80200...`. Because all methods in PhpVersion asking about this number always ask about the minor version, ending in `00`. Meaning that if `PHP_VERSION_ID` is 80210, the `10` at the end would have to be trimmed and exchanged for `00` to correctly answer.

Does this sound compelling to you?

Reply:
yeah I can follow this train :).

> when rules take the php-version from the `Scope`, we also need to think how we can/want control the `PhpVersion` in tests, so we can test `PhpVersion` dependent behaviour

this is still a open question though. how to adjust existing tests so we can force the php version?

wdyt about adding `RuleTestCase->getPhpVersions()` with a default impl and handle it similar to `getCollectors`?

Reply:
I don't think we need anything like that. If we want to test behaviour for a specific PHP version we will just write the condition `if (PHP_VERSION_ID > ...)` in the tested code.